rfd/0xxx-partial-cache-healthiness.md

```diff
@@ -0,0 +1,322 @@
+it back to unhealthy state. Every time it becomes healthy again, it's considered a new *generation* of cache. Since every
+time it reconnects to the Auth service, it might be of different versions, the list of rejected kinds should be considered
+specific to the current generation and refreshed during each re-initialization.
+
+Since cache can also act as an event source, it should propagate the rejected status if a client tries to use the cache
+to subscribe to events of an object kind that was rejected to the cache itself. Such client will also receive a
+list of rejected kinds on the first `OpInit` event or a failure depending on the supplied AllowRejectedKinds flag. 
+
+
+### Changes to the Events interface
+
+Because the `Events` interface is central to the caching subsystem, it will require some changes. Also, because it has
+so many implementations and is used in so many contexts, some rules need to be established and all the implementations
+must be updated to behave in a consistent way.
+
+None of the proposed changes to the interface are direct though. They all happen through related types and behaviour.
+
+
+#### types.Watch
+```diff
+type Watch struct {
+	Name string
+	Kinds []WatchKind
+	QueueSize int
+	MetricComponent string
++
++	// AllowRejectedKinds enables partial success mode in which the request won't fail if some of the requested Kinds
++	// are not available for watching.
++	AllowRejectedKinds bool
++
++	// KindsRejectedInFlight contains a list of kinds rejected by upstream implementations in a chain of NewWatcher()
++	// calls. Must be empty if AllowRejectedKinds is not set. A kind cannot be on both Kinds and KindsRejectedInFlight
++	// at the same time.
++	KindsRejectedInFlight []WatchKind
+}
+```
+
+The new field `AllowRejectedKinds` should be added to preserve backwards compatibility when a request for an event stream
+crosses service boundary. If an old downstream service unaware of partial success mode connects to an upgraded auth server,
+it won't set this flag. In that case it should get an error if one of the kinds requested for watching is unavailable.
+
+`KindsRejectedInFlight` should be added to accommodate wrapper implementations of `Events` which perform validations
+and call the inner implementation. If `AllowRejectedKinds` is set and the wrapper determines that some of the requested
+kinds cannot be watched, they will be moved from Kinds to this new field and passed to the inner implementation this way.
+
+#### types.WatchStatus
+
+The `Event` struct has fields for a type and an associated resource. Currently, events with type `OpInit` have
+the associated resource field set to `nil`. We should add a new resource type that can be attached to `OpInit` events
+and carry the list of resource kinds what weren't available for watching:
+
+```protobuf
+message WatchStatusV1 {
+  string Kind = 1;
+  string SubKind = 2;
+  string Version = 3;
+  Metadata Metadata = 4;
+  WatchStatusSpecV1 Spec = 5;
+}
```

Comment:
It's true that `WatchStatus` is not a real resource, but to be assignable to `event.Resource` it still needs to implement the `Resource` interface and have methods like`GetKind()`, `GetMetadata()`, etc. We could omit the resource header in protobuf and just implement those getters as returning string literals, e.g.

```go
func (m *WatchStatusV1) GetKind() string {
	return "watch_status"
}
```

Personally, I would rather pretend that it's a real resource for consistency and in case we decide to pass more information via the headers in the future.